motionplan/rrtStarConnect.go

```diff
@@ -176,64 +235,208 @@ func (mp *rrtStarConnectMotionPlanner) rrtBackgroundRunner(ctx context.Context,
+	copy(qstep, mp.algOpts.qstep)
+	doubled := false
+
+	oldNear := near
+	// This should iterate until one of the following conditions:
+	// 1) we have reached the target
+	// 2) the request is cancelled/times out
+	// 3) we are no longer approaching the target and our "best" node is further away than the previous best
+	// 4) further iterations change our best node by close-to-zero amounts
+	// 5) we have iterated more than maxExtendIter times
+	for i := 0; i < maxExtendIter; i++ {
+		select {
+		case <-ctx.Done():
+			mchan <- oldNear
+			return
+		default:
+		}
+
+		// iterate over the k nearest neighbors and find the minimum cost to connect the target node to the tree
+		neighbors := kNearestNeighbors(mp.planOpts, rrtMap, &basicNode{q: target.Q()}, mp.algOpts.NeighborhoodSize)
+
+		dist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: near.Q(), EndConfiguration: target.Q()})
+		oldDist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: oldNear.Q(), EndConfiguration: target.Q()})
+		switch {
+		case dist < mp.algOpts.JointSolveDist:
+			mchan <- near
+			return
+		case dist > oldDist:
+			mchan <- oldNear
+			return
+		}
+
+		oldNear = near
+		newNear := make([]referenceframe.Input, 0, len(near.Q()))
+
+		// alter near to be closer to target
+		for j, nearInput := range near.Q() {
+			if nearInput.Value == target.Q()[j].Value {
+				newNear = append(newNear, nearInput)
+			} else {
+				v1, v2 := nearInput.Value, target.Q()[j].Value
+				newVal := math.Min(qstep[j], math.Abs(v2-v1))
+				// get correct sign
+				newVal *= (v2 - v1) / math.Abs(v2-v1)
+				newNear = append(newNear, referenceframe.Input{nearInput.Value + newVal})
+			}
+		}
+		// Check whether newNear meets constraints, and if not, update it to a configuration that does meet constraints (or nil)
+		newNear = mp.constrainNear(ctx, randseed, oldNear.Q(), newNear)
+
+		if newNear != nil {
+			nearDist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: oldNear.Q(), EndConfiguration: newNear})
+			if nearDist < math.Pow(mp.algOpts.JointSolveDist, 3) {
+				if !doubled {
+					doubled = true
+					// Check if doubling qstep will allow escape from the identical configuration
+					// If not, we terminate and return.
+					// If so, qstep will be reset to its original value after the rescue.
+					for i, q := range qstep {
+						qstep[i] = q * 2.0
+					}
+					continue
+				} else {
+					// We've arrived back at very nearly the same configuration again; stop solving and send back oldNear.
+					// Do not add the near-identical configuration to the RRT map
+					mchan <- oldNear
+					return
+				}
+			}
+			if doubled {
+				copy(qstep, mp.algOpts.qstep)
+				doubled = false
+			}
+			// constrainNear will ensure path between oldNear and newNear satisfies constraints along the way
+			near = &basicNode{q: newNear, cost: neighbors[0].node.Cost() + neighbors[0].dist}
+			rrtMap[near] = oldNear
+
+			// rewire the tree
+			for i, thisNeighbor := range neighbors {
```

Comment:
Wouldn't neighbors need to be created based on `newNear`, as that's what is actually being added to the tree, not `target`?